#------------- EKF parameters ----------------
#
# Defines parameters used by the Extended Kalman Filter:
# - Sensor bias
# - Covariance matrices
#
# ----------------------------------------------------

## general settings for the EKF
is_simulation: false # 1=runs in simulation (SIL), 0=runs on drone
use_gps: true #!!only read if is_simulation=0!!:0=use optitrack, 1=use px4 gps
use_magnetometer: true #!!only read if is_simulation=0!!:0=do not use px4 magnetometer, 1=use px4 magnetometer
use_barometer: true #!!only read if is_simulation=0!!:0=do not use px4 barometer, 1=use px4 barometer
predict_on_idle: true #!!only read if is_simulation=0!!:1:navigation runs on idle state, 0:navigation does not run on idle state
dt_state_publish: 0.05 # [s] time delay between each publish of EKF state estimation (/kalman_rocket_state message)

imu_calibration_number_of_samples: 2000 #!!only used if is_simulation=0!!:1: number of samples needed to calibrate imu


## Sensor bias

gyro_bias_x: 0.0 # [rad.s^-1]
gyro_bias_y: 0.0 # [rad.s^-1]
gyro_bias_z: 0.0 # [rad.s^-1]

acc_bias_x: 0.0 # [m.s^-2]
acc_bias_y: 0.0 # [m.s^-2]
acc_bias_z: 0.0 # [m.s^-2]

mag_bias_x: 0.0 # [T]
mag_bias_y: 0.0 # [T]
mag_bias_z: 0.0 # [T]

mag_declination: -0.04712389 #(value for lausanne in may 2022, ! negative) #[rad] #magnetic declination at launch

# magnetometer calibration matrices: A and B such that: m_calibrated = A*(m_uncalibrated - b)
A_mag_00: 0.9825
A_mag_01: 0.0130
A_mag_02: 0.0072
A_mag_10: 0.0130
A_mag_11: 0.9890
A_mag_12: 0.0106
A_mag_20: 0.0072
A_mag_21: 0.0106
A_mag_22: 1.0295

b_mag_0: 0.0000000717
b_mag_1: -0.0000000764
b_mag_2: 0.0000004612

### covariance matrices

##  Definition of W (process covariance matrix Q = GWG' with G = df/dw (non-additive noise formulation))

#W_gyro_x: 0.00002182
#W_gyro_y: 0.00002528
#W_gyro_z: 0.00000384
#W_acc_x: 0.0053
#W_acc_y: 0.0072
#W_acc_z: 0.0052
#W_acc_bias_x:  0.1
#W_acc_bias_y:  0.1
#W_acc_bias_z:  0.1
#W_gyro_bias_x: 0.1
#W_gyro_bias_y: 0.1
#W_gyro_bias_z: 0.1
#W_baro_bias: 0.000001
#W_mag_x: 1
#W_mag_y: 1
#W_mag_z: 1
#W_mag_bias_x: 0.1
#W_mag_bias_y: 0.1
#W_mag_bias_z: 0.1
W_gyro_x: 0.00002182
W_gyro_y: 0.00002528
W_gyro_z: 0.00000384
W_acc_x: 0.0053
W_acc_y: 0.0072
W_acc_z: 0.0052
W_acc_bias_x:  0.00001
W_acc_bias_y:  0.00001
W_acc_bias_z:  0.00001
W_gyro_bias_x: 0.0000000001
W_gyro_bias_y: 0.0000000001
W_gyro_bias_z: 0.0000000001
W_baro_bias: 0.000001
W_mag_x: 0.01
W_mag_y: 0.01
W_mag_z: 0.01
W_mag_bias_x: 0.001
W_mag_bias_y: 0.001
W_mag_bias_z: 0.001


## measurement update covariances (R) / sensor covarience matrix
## ONLY THE DIAGONALS

R_gyro_x: 0.00002182 #gyroscope covariance [rad^2.s^-2]
R_gyro_y: 0.00002528  #gyroscope covariance [rad^2.s^-2]
R_gyro_z: 0.00000384 #gyroscope covariance [rad^2.s^-2]

R_mag_x: 0.0078   #magnetometer covariance normalised mag vec [-]
R_mag_y: 0.0078   #magnetometer covariance normalised mag vec [-]
R_mag_z: 0.0078   #magnetometer covariance normalised mag vec [-]

R_baro: 0.04 #barometer covariance [m^2]

R_gps_x: 0.64 #gps covariance [m^2] # we get this information from the gps system when not in simulation
R_gps_y: 0.64 #gps covariance [m^2] # we get this information from the gps system when not in simulation
R_gps_z: 0.64 #gps covariance [m^2] # we get this information from the gps system when not in simulation

R_optitrack_x: 0.000000001 #optitrack covariance [m^2]
R_optitrack_y: 0.000000001 #optitrack covariance [m^2]
R_optitrack_z: 0.000000001 #optitrack covariance [m^2]


## fake gps parameters (this is only used in simulation)
fake_gps_freq: 20
fake_gps_variance: 0.8
